src/main/kotlin/org/rust/ide/presentation/TypeRendering.kt

```diff
@@ -266,15 +257,30 @@ private data class TypeRenderer(
     private fun formatBoundElementGenerics(
         boundElement: BoundElement<RsGenericDeclaration>,
         render: (Ty) -> String
+    ): List<String> = formatGenerics(boundElement.element, boundElement.subst, render)
+
+    private fun formatGenerics(
+        declaration: RsGenericDeclaration,
+        subst: Substitution,
+        render: (Ty) -> String
     ): List<String> {
-        val tySubst = boundElement.element.typeParameters.map { render(boundElement.subst[it] ?: TyUnknown) }
-        val regionSubst = if (includeLifetimeArguments) {
-            boundElement.element.lifetimeParameters.map { render(boundElement.subst[it] ?: ReUnknown) }
-        } else {
-            emptyList()
-        }
-        val constSubst = boundElement.element.constParameters.map { render(boundElement.subst[it] ?: CtUnknown) }
-        return regionSubst + tySubst + constSubst
+        var nonDefaultTyFound = false
+        return declaration.genericParameters
+            .asReversed()
+            .filter { parameter ->
+                if (!skipUnchangedDefaultTypeArguments || parameter !is RsTypeParameter) return@filter true
+                nonDefaultTyFound = nonDefaultTyFound || parameter.typeReference?.type != subst[parameter]
+                nonDefaultTyFound
```

Comment:
Unfortunately, it doesn't work as expected.
For example, for the following code (yep, I know, that it's not allowed yet but anyway):
```rust
#![feature(const_generics)]

struct S<T=u32, const B: usize>([T; B]);

fn main() {
    let s = S([1u32]);
    s;
}
```

if you try to extract `s` into function via the corresponding refactoring, you get
```rust
#![feature(const_generics)]

struct S<T=u32, const B: usize>([T; B]);

fn main() {
    let s = S([1u32]);
    foo(s);
}

fn foo(s: S<1>) -> S<1> {
    s
}
```
but I expect
```rust
#![feature(const_generics)]

struct S<T=u32, const B: usize>([T; B]);

fn main() {
    let s = S([1u32]);
    foo(s);
}

fn foo(s: S<u32, 1>) -> S<u32, 1> {
    s
}
```

You need always change `nonDefaultTyFound` (BYW not the best name since we can find const parameter) to `true` if `parameter` is not `RsTypeParameter`

Reply:
Oh, it's even checked in your tests but the corresponding test is wrong